### PR TITLE
Add base implementation for realtime communication

### PR DESCRIPTION
**Pull request by mrdrogdrog**
### Component/Part
Realtime communication

### Description
This PR adds the base implementation for realtime communication via yjs.

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

### Related Issue(s)

https://github.com/hedgedoc/react-client/pull/2118

### CODE REVIEW

**Review by DerMolly — CHANGES_REQUESTED** [review]

File: src/realtime/websocket/extract-note-id-from-path.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * Extracts the note id from the given realtime URL path.
+ *
+ * @param realtimePath The path from which the note id should be extracted
+ * @return The extracted note id
+ * @throws Error if the given string isn't a valid realtime URL path
+ */
+export function extractNoteIdFromPath(realtimePath: string): string {
+  return Optional.ofNullable(new URL(realtimePath, 'https://example.org'))
```
[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply src/realtime/websocket/extract-note-id-from-path.ts:16]: You can't use the URL constructor with only a relative part. It needs either an origin in the path itself or as the second argument. (We also use that in the react client to create absolute links in the renderer.)
But because we actually don't care about the domain or protocol I used "example.org".

Why did I choose to use the URL constructor anyway? Because it knows EVERYTHING about parsing an URL. Every corner case. We need one parameter, but what if we decide to pass other parameters? With this approach we don't have to change the logic here.

File: src/notes/notes.service.ts [hunk 2]
```diff
@@ -117,27 +118,13 @@ export class NotesService {
...
-    return (await this.getLatestRevision(note)).content;
-  }
-
-  /**
-   * @async
-   * Get the first revision of the note.
-   * @param {Note} note - the note to use
-   * @return {Revision} the first revision of the note
-   */
-  async getLatestRevision(note: Note): Promise<Revision> {
```
Comment (davidmehren) [inline src/notes/notes.service.ts:129]: Why was this deleted?

**Review by davidmehren — CHANGES_REQUESTED** [review]
I'm not really sure if this is already ready for review, but I had a look anyway.

Refactoring work (like deleting references to `NotesService` from the permissions service/module) was done without any comment at various places and belongs into separate commits.

File: package.json [hunk 3]
```diff
@@ -92,10 +105,12 @@
     "http-proxy-middleware": "2.0.6",
     "jest": "28.1.1",
     "mocked-env": "1.3.5",
+    "optional-js": "2.3.0",
```
[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply package.json:108]: > unmaintained
Or: Because this lib ist just feature complete and doesn't have any dependencies ist is... stable! :scream:

File: package.json [hunk 4]
```diff
@@ -92,10 +105,12 @@
...
     "prettier": "2.6.2",
     "source-map-support": "0.5.21",
     "supertest": "6.2.3",
     "ts-jest": "28.0.4",
+    "ts-mockery": "1.2.0",
```
[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply package.json:113]: It doesn't replace any jest stuff. This lib provides type safe mocks.

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (SISheogorath) [reply package.json:108]: looking at the maintainers github profile, I'm not sure how maintained it can be: Literally no activity in 2021 and 2022. In 2020 there was one busy Wednesday.

His website seems also gone. So… feature complete but unmaintained?

File: src/notes/notes.service.ts (hunk 2 above)
Comment (mrdrogdrog) [reply src/notes/notes.service.ts:129]: Because it was a unnecessary function that was only used in places that also import the revisions service. And also this method provoked a lot of cycling dependencies.

File: src/realtime/websocket/extract-note-id-from-path.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+/**
+ * Extracts the note id from the given realtime URL path.
+ *
+ * @param realtimePath The path from which the note id should be extracted
+ * @return The extracted note id
+ * @throws Error if the given string isn't a valid realtime URL path
+ */
+export function extractNoteIdFromPath(realtimePath: string): string {
```
Comment (InnayTool) [inline src/realtime/websocket/extract-note-id-from-path.ts:14]: Why is this file named `extract-note-id-from-path.ts` and not `util.ts`? And why is this function not part of `websocket.gateway.ts`?

File: src/realtime/websocket/extract-note-id-from-path.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  const url = new URL(realtimePath, 'https://example.org');
```
[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply src/realtime/websocket/extract-note-id-from-path.ts:15]: Has already been answered in the review of @dermolly

File: src/realtime/realtime-note/realtime-note.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    this.logger.log(`New client connected`);
+    this.clients.add(client);
+  }
+
+  /**
+   * Disconnects the given websocket client while cleaning-up if it was the last user in the realtime note.
+   *
+   * @param {WebSocket} client The websocket client that disconnects.
+   */
+  public removeClient(client: WebsocketConnection): void {
```
[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply src/realtime/realtime-note/realtime-note.ts:53]: i would rather rename the other method because it's not disconnecting.

File: src/realtime/websocket/extract-note-id-from-path.ts (hunk 5 above)
Comment (mrdrogdrog) [reply src/realtime/websocket/extract-note-id-from-path.ts:14]: Why should it be called util? It exports one function.

Easier to test and less cluttering in the gateway file.

File: src/realtime/websocket/extract-note-id-from-path.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (InnayTool) [reply src/realtime/websocket/extract-note-id-from-path.ts:15]: Yes I see, but we should add an comment here.

File: src/realtime/realtime-note/realtime-note.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (InnayTool) [reply src/realtime/realtime-note/realtime-note.ts:53]: Yes, also possible.

File: src/realtime/websocket/extract-note-id-from-path.ts (hunk 5 above)
Comment (InnayTool) [reply src/realtime/websocket/extract-note-id-from-path.ts:14]: I think it's inconsistent. There are many other static functions in the gateway file which are not extracted.

**Review by DerMolly — APPROVED** [review]

**Review by InnayTool — APPROVED** [review]

File: src/realtime/websocket/utils/extract-note-id-from-request-url.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * @param request The request whose URL should be extracted
+ * @return The extracted note id
+ * @throws Error if the given string isn't a valid realtime URL path
+ */
+export function extractNoteIdFromRequestUrl(request: IncomingMessage): string {
+  if (request.url === undefined) {
+    throw new Error('No URL found in request');
+  }
+  const url = new URL(request.url, 'https://example.org'); //Need a valid origin for the URL constructor to work.
```
Comment (ErikMichelson) [inline src/realtime/websocket/utils/extract-note-id-from-request-url.ts:19]: Improvement suggestion for the comment:

```suggestion
  const url = new URL(request.url, 'https://example.org');
  // A valid domain name is needed for the URL constructor, although not being used here.
  // The example.org domain should be safe to use according to RFC 6761.
```

File: src/realtime/websocket/websocket.gateway.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+      const user = await this.findUserByRequestSession(request);
+      const note = await this.noteService.getNoteByIdOrAlias(
+        extractNoteIdFromRequestUrl(request),
+      );
+
+      if (!(await this.permissionsService.mayRead(user, note))) {
+        this.logger.log(
+          `Access denied to note '${note.id}' for user '${user.username}'`,
+          'handleConnection',
+        );
```
Comment (ErikMichelson) [inline src/realtime/websocket/websocket.gateway.ts:59]: We had a TODO here, that we should inform the client about the failure, so that the client does not try to reconnect the closed connection.
Do we want to re-add this TODO here?

**Review by ErikMichelson — APPROVED** [review]

**Review by davidmehren — CHANGES_REQUESTED** [review]

**Review by davidmehren — APPROVED** [review]
